examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,417 @@
+
+"""
+# Setting up the Embedding generator model
+
+Our Siamese Network will generate embeddings for each one of the images of the
+triplet. To do this, we will use a pre-trained ResNet50 model on ImageNet and
+connect a few `Dense` layers to it so we have space to learn to separate these
+embeddings.
+
+We will freeze the weights of all the layers of the model up until
+`conv5_block1_out`. This is important so we don't mess with the weights that
+the model already learned. We are going to leave the bottom few layers open so
+we can fine tune those weights during training.
+"""
+
+base_cnn = applications.ResNet50(
+    weights="imagenet", input_shape=target_shape + (3,), include_top=False
+)
+
+flatten = layers.Flatten()(base_cnn.output)
+dense1 = layers.Dense(512, activation="relu")(flatten)
+dense1 = layers.BatchNormalization()(dense1)
+dense2 = layers.Dense(256, activation="relu")(dense1)
+dense2 = layers.BatchNormalization()(dense2)
+output = layers.Dense(256)(dense2)
+
+embedding = Model(base_cnn.input, output, name="Embedding")
+
+trainable = False
+for layer in base_cnn.layers:
+    if layer.name == "conv5_block1_out":
+        trainable = True
+    layer.trainable = trainable
+
+"""
+# Setting up the Siamese Network model
+
+The Siamese network will receive each one of the triplet images as an input,
+generate the embeddings, and output the distance between the anchor and the
+positive embedding, and the distance between the anchor and the negative
+embedding.
+
+To compute the distance, we can use a custom layer `DistanceLayer` that
+returns both values as a tuple.
+"""
+
+
+class DistanceLayer(layers.Layer):
+    """
+    This layer is responsible for computing the distance between the anchor
+    embedding and the positive embedding, and the anchor embedding and the
+    negative embedding.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def call(self, anchor, positive, negative):
+        ap_distance = tf.reduce_sum(tf.square(anchor - positive), -1)
+        an_distance = tf.reduce_sum(tf.square(anchor - negative), -1)
+        return (ap_distance, an_distance)
+
+
+anchor_input = layers.Input(name="anchor", shape=target_shape + (3,))
+positive_input = layers.Input(name="positive", shape=target_shape + (3,))
+negative_input = layers.Input(name="negative", shape=target_shape + (3,))
+
+distances = DistanceLayer()(
+    embedding(preprocess_input(anchor_input)),
+    embedding(preprocess_input(positive_input)),
+    embedding(preprocess_input(negative_input)),
+)
+
+siamese_network = Model(
+    inputs=[anchor_input, positive_input, negative_input], outputs=distances,
+)
+
+"""
+# Putting everything together
+
+We now need to implement a model with custom training loop so we can compute
+the Triplet Loss using the three embeddings produced by the Siamese network.
+
+Here is the definition of Triplet Loss implemented on this example:
+`L(A,P,N) = max(||f(A)-f(P)||**2 - ||f(A)-f(N)||**2 + alpha, 0)`
```

Comment:
Suggestion:

```suggestion
We now need to implement a model with custom training loop so we can compute
the triplet loss using the three embeddings produced by the Siamese network.
```

(the loss is defined in the `class SiameseModel()`.. alternatively, delete the formula there and include it here )